utilities/transactions/lock/range/range_tree/range_tree_lock_tracker.cc

```diff
@@ -0,0 +1,65 @@
+//
+// A lock tracker to be used together with RangeTreeLockManager
+//
+#ifndef ROCKSDB_LITE
+#ifndef OS_WIN
+
+#include "range_tree_lock_tracker.h"
+
+#include "range_tree_lock_manager.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+RangeLockList *RangeTreeLockTracker::getOrCreateList() {
+  RangeLockList *res;
+  if ((res = getList())) return res;
+
+  // Doesn't exist, create
+  range_list.reset(new RangeLockList());
+  return getList();
+}
+
+void RangeTreeLockTracker::Track(const PointLockRequest &lock_req) {
+  DBT key_dbt;
+  std::string key;
+  serialize_endpoint(Endpoint(lock_req.key, false), &key);
+  toku_fill_dbt(&key_dbt, key.data(), key.size());
+  RangeLockList *rl = getOrCreateList();
+  rl->append(lock_req.column_family_id, &key_dbt, &key_dbt);
+}
+
+void RangeTreeLockTracker::Track(const RangeLockRequest &lock_req) {
+  DBT start_dbt, end_dbt;
+  std::string start_key, end_key;
+
+  serialize_endpoint(lock_req.start_endp, &start_key);
+  serialize_endpoint(lock_req.end_endp, &end_key);
+
+  toku_fill_dbt(&start_dbt, start_key.data(), start_key.size());
+  toku_fill_dbt(&end_dbt, end_key.data(), end_key.size());
+
+  RangeLockList *rl = getOrCreateList();
+  rl->append(lock_req.column_family_id, &start_dbt, &end_dbt);
```

Comment:
`start_dbt` and `end_dbt` are destroyed after `Track`. Here we pass their addresses to `rl`, is this safe since after `Track`, the addresses are no longer meaningful?

Reply:
It is safe, because the code inside RangeLockList::append() will copy the contents of DBT structures. It will not retain pointers to them.
The copying happens here:
```
  #0  toku::range_buffer::append_range (...) at utilities/transactions/lock/range/range_tree/lib/locktree/range_buffer.cc:235
  #1  0x0000555555aeea72 in toku::range_buffer::append (...) at utilities/transactions/lock/range/range_tree/lib/locktree/range_buffer.cc:205
  #2  0x00005555559e614c in rocksdb::RangeLockList::Append (...) at utilities/transactions/lock/range/range_tree/range_tree_lock_tracker.cc:76
  #3  0x00005555559e5e77 in rocksdb::RangeTreeLockTracker::Track (...) at utilities/transactions/lock/range/range_tree/range_tree_lock_tracker.cc:41
  #4  0x0000555555af85eb in rocksdb::PessimisticTransaction::GetRangeLock (...) at utilities/transactions/pessimistic_transaction.cc:682
```
https://github.com/facebook/rocksdb/blob/master/utilities/transactions/lock/range/range_tree/lib/locktree/range_buffer.cc#L235